tests/unit_tests/ringdb.cpp

```diff
@@ -0,0 +1,158 @@
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <boost/filesystem.hpp>
+
+#include "gtest/gtest.h"
+
+#include "string_tools.h"
+#include "crypto/crypto.h"
+#include "crypto/random.h"
+#include "crypto/chacha.h"
+#include "wallet/ringdb.h"
+
+static crypto::chacha_key generate_chacha_key()
+{
+  uint8_t key[CHACHA_KEY_SIZE];
+  crypto::rand(CHACHA_KEY_SIZE, key);
+  crypto::chacha_key chacha_key;
+  memcpy(&chacha_key, key, CHACHA_KEY_SIZE);
+  return chacha_key;
+}
+
+static crypto::key_image generate_key_image()
+{
+  return crypto::rand<crypto::key_image>();
+}
+
+static crypto::public_key generate_output()
+{
+  return crypto::rand<crypto::public_key>();
+}
+
+
+static const crypto::chacha_key KEY_1 = generate_chacha_key();
+static const crypto::chacha_key KEY_2 = generate_chacha_key();
+static const crypto::key_image KEY_IMAGE_1 = generate_key_image();
+static const crypto::public_key OUTPUT_1 = generate_output();
+static const crypto::public_key OUTPUT_2 = generate_output();
+
+class RingDB: public tools::ringdb
+{
+public:
+  RingDB(const char *genesis = ""): tools::ringdb(make_filename(), genesis) { }
+  ~RingDB() { boost::filesystem::remove_all(filename); free(filename); }
+
+private:
+  std::string make_filename()
+  {
+    boost::filesystem::path path = tools::get_default_data_dir();
+    path /= "fake";
+    path /= "ringdb-test-XXXXXX";
+    filename = strdup(path.string().c_str());
+    EXPECT_TRUE(mkdtemp(filename)!= NULL);
```

Comment:
I'm not familiar with MinGW environment, but `mkdtemp` seems to be unavailable for MinGW-w64 (https://sourceforge.net/p/mingw-w64/code/HEAD/tree/stable/v3.x/mingw-w64-headers/crt/stdlib.h):

```
/monero/tests/unit_tests/ringdb.cpp:83:17: error: 'mkdtemp' was not declared in this scope
     EXPECT_TRUE(mkdtemp(filename)!= NULL);
```

Adding `#include <stdlib.h>` didn't make any difference.

It does seem to exist in the original MinGW (https://sourceforge.net/p/mingw/mingw-org-wsl/ci/mingwrt-3.21.1-release/tree/mingwrt/include/stdlib.h), so this is maybe simply due to the lack of support?